dynamo-adapter.js

```diff
@@ -0,0 +1,145 @@
+const {promisify} = require('util');
+const {DynamoTable} = require('./slim');
+
+const isCallback = (fn) => typeof fn === 'function';
+
+const validateCall = (methodName, args, minLength) => {
+  if (args.length < minLength)
+    throw new Error(`Incorrect args supplied to method ${methodName}: expected: ${minLength} got: ${args.length}`);
+}
+
+/**
+  Provides an adapter for the DynamoTable class to allow for promisified
+  api usage when omitting the cb parameter.
+*/
+class DynamoAdapter {
+  constructor(name, options) {
+    this._delegate = new DynamoTable(name, options);
+    this.hooks = options.hooks || {};
+
+    // Delegate externally accessed properties
+    Object.keys(this._delegate).forEach(key => {
+      this[key] = this._delegate[key];
+    });
+
+    // NOTE: Rebind all methods so they can be passed without the class instance
+    // Since we only instantiate once per table and maintain the instances,
+    // the performance should be a big concern
+    Object.getOwnPropertyNames(DynamoAdapter.prototype).forEach(key => {
+      if (typeof this[key] === 'function') this[key] = this[key].bind(this);
+    });
+
+    // NOTE: This function needs to be rebound to the delegate for its use case
+    this['mapToDb'] = this._delegate.mapToDb.bind(this._delegate);
+  }
+
+  // Proxies execution to delegate
+  _call(methodName, minLength, ...args) {
+    validateCall(methodName, args, minLength);
+
+    const cb = args[args.length - 1];
+
+    let method = this._delegate[methodName];
+
+    // TODO: No support for async 'hook'
+    this._runHook(methodName);
+
+    if (!isCallback(cb)) method = promisify(method);
+
+    return method.apply(this._delegate, args);
+  }
+
+  _runHook(methodName) {
+    const hook = this.hooks[methodName];
+
+    return hook ? hook() : null;
+  }
+
+  // Allows additional hook for execution BEFORE function
+  addHook(methodName, fn) {
+    this.hooks[methodName] = fn;
+  }
+
+  // Provides instance copy functionality
+  copy(options) {
+    const overrides = Object.assign({}, this, options);
+
+    return new DynamoAdapter(this.name, overrides);
+  }
+
+  // Accepts minimum 1 parameter
+  addNew(...args) {
```

Comment:
seems a bit boilerplate from this point. Alternatively:
```
Object.assing(this, {
    count: this_.call.bind(this, 'count', 0),
    get: this_.call.bind(this, 'get', 1),
    /// ...bla
```
or even
```
minParams = {
    count: 0,
    get: 1,
    //... not as explicit 
```

Reply:
Well the first suggestion is (sort of) the same as current just without the es6 methods which sadly don't have the nicer class properties syntax. Second was something i did consider before, originally didn't do that as didn't want to do it in constructor but later changed to do the binding there anyway so probably wouldn't have mattered.

I think first is best if had property syntax such as:

```javascript
count = (...args) => this._call.bind(this, 'count', 0, ...args),
get = (...args) => this._call.bind(this, 'get', 1, ...args),
```

Imo it doesn't really need to change again for this as its just syntax and the assign is not identical to the specified methods. Maybe if comment was from back when previously changing then would consider more but now all green I dont think its important.

Reply:
good to me